clash_lib/src/app/router/rules/geodata/mod.rs

```diff
@@ -0,0 +1,182 @@
+use crate::app::router::RuleMatcher;
+use crate::session::Session;
+use std::fmt::{Display, Formatter};
+
+use crate::app::router::rules::geodata::attribute::{AndAttrMatcher, AttrMatcher};
+use crate::app::router::rules::geodata::matcher_group::{DomainGroupMatcher, SuccinctMatcherGroup};
+use crate::common::geodata::GeoData;
+
+mod attribute;
+mod matcher_group;
+mod str_matcher;
+
+// if country_code is empty, return None
+// or will return the parsed **real** code and the attr list and if the code is negated
+fn parse(country_code: &str) -> Option<(bool, String, Box<dyn AttrMatcher>)> {
+    let country_code = country_code.trim().to_lowercase();
+    let mut country_code = country_code.as_str();
+    let mut not = false;
+
+    if country_code.is_empty() {
+        return None;
+    }
+
+    if country_code.as_bytes()[0] == b'!' {
+        not = true;
+        country_code = &country_code[1..];
+    }
+    let parts = country_code.split('@').collect::<Vec<&str>>();
+    let code = parts[0].to_owned();
+    let attrs = if parts.len() > 1 {
+        parts[1].split(',').map(|x| x.to_owned()).collect()
+    } else {
+        Vec::new()
+    };
+    let attr_matcher = Box::new(AndAttrMatcher::from(attrs)) as _;
+
+    Some((not, code, attr_matcher))
+}
+
+pub struct GeoSiteMatcher {
+    pub country_code: String,
+    pub target: String,
+    pub matcher: Box<dyn DomainGroupMatcher>,
+}
+
+impl GeoSiteMatcher {
+    pub fn new(country_code: String, target: String, loader: &GeoData) -> anyhow::Result<Self> {
+        let (not, code, attr_matcher) = parse(&country_code).unwrap();
```

Comment:
we probably shouldn't crash or print en error if the country_code is empty

Reply:
ok